### DISCUSSION TITLE
Question around counter-intuitive performance with Expect Header

### DISCUSSION THREAD

**Discussion by tailrecur** [curl/curl#6635] [Q&A]
Thanks a lot for `curl` !

From my understanding of the `Expect` Header, its presence causes an additional roundtrip to the server which leads to an additional delay (upto`CURLOPT_EXPECT_100_TIMEOUT_MS` configuration). As I understand, disabling the `Expect` header usage should reduce latency since we avoid the client waiting before sending the request body of a `POST` request.

I'm building a testing tool to validate the effect of disabling the `Expect` header on total latency and hopefully [add support](https://github.com/sagebind/isahc/issues/303) for this in a Rust HTTP client library.

However, when I configured `libcurl` to not use `Expect` by setting a header `Expect:`, I observe that the latency of the actual POST request becomes worse. [This comment](https://github.com/sagebind/isahc/issues/303#issuecomment-782175587) details the observed behaviour with multiple requests to the same url with a re-used connection alternating between enabling and disabling the `Expect` header.

Looking through the `curl` source code, I found the [following intriguing comment](https://github.com/curl/curl/blob/master/lib/http.c#L1671):
```
/* if not doing HTTP 1.0 or version 2, or disabled explicitly, we add an
       Expect: 100-continue to the headers which actually speeds up post
       operations (as there is one packet coming back from the web server) */
```

I tried git bisecting the commit and found that this comment was added by @bagder [back in 2001](https://github.com/curl/curl/commit/08655d8d5d0ea980227096366c231693198e61d6#diff-14c8235d370910f1781de1a00ed1957ec948e84e88fdc52db87bd4eb7c7167ffR726) as part of a larger changeset.

I don't understand what the part `as there is one packet coming back from the web server` means and how that helps improve performance. Googling for it did not offer much either. It would be great if someone can explain this behaviour in more detail. 

Thanks a lot!

**Comment by bagder** [comment]
> From my understanding of the `Expect` Header, its presence causes an additional roundtrip to the server

If and only if the server returns a "100 Continue" to the request, then yes.

If the server instead returns a non-200 response, you might have gained a lot more as you didn't have to send any data and you can keep the connection alive.

> disabling the Expect header usage should reduce latency since we avoid the client waiting before sending the request body of a POST request.

It certainly *might* reduce the latency, especially if the server is stupid and doesn't support the `Expect:` header as that would otherwise make libcurl wait 1000 ms and then continue anyway, ending up just wasting that time.

I can't recall what I meant with that comment.

Two details that have changed since the ancient days include that curl now a) sets `TCP_NODELAY` by default, which it didn't back in the day and that changes behavior especially in reused connections and b) uses `Expect:` on much larger content by default.

It would help if you could qualify what exactly "it becomes worse" means. I mean, where/how does it change and in which end.

**Reply by tailrecur** [reply]
@bagder Thanks for the detailed reply. 
I now understand better the implications of using the `Expect` header.

I disabled the `Expect` header but the total latency for a successful POST request was still roughly twice the roundtrip time. To dig deeper, I started a `tcpdump` and then made the HTTP request with `libcurl` in verbose mode. I've pasted an excerpt of the output after changing the url and body for anonymisation:

### libcurl:

<pre>
[2021-02-27T08:37:13.582602Z DEBUG isahc::handler] Found bundle for host remote-server.com: 0x7fae6ed04c50 [serially]
[2021-02-27T08:37:13.582770Z DEBUG isahc::handler] Re-using existing connection! (#0) with host remote-server.com
[2021-02-27T08:37:13.582863Z DEBUG isahc::handler] Connected to remote-server.com (52.74.146.247) port 443 (#0)
<b>[2021-02-27T08:37:13.583094Z TRACE isahc::handler] >> POST /example-uri HTTP/1.1\r\nHost: remote-server.com\r\nAccept: */*\r\nAccept-Encoding: deflate, gzip\r\ncontent-type: application/json\r\nuser-agent: carbine/0.0.1\r\nconnection: keep-alive\r\nContent-Length: 636\r\n\r\n
[2021-02-27T08:37:13.583603Z TRACE isahc::handler] >> {"body": "elided request body"}\n
[2021-02-27T08:37:13.584035Z DEBUG isahc::handler] We are completely uploaded and fine</b>
[2021-02-27T08:37:14.298139Z DEBUG isahc::handler] Mark bundle as not supporting multiuse
[2021-02-27T08:37:14.298199Z TRACE isahc::handler] << HTTP/1.1 200 OK\r\n
[2021-02-27T08:37:14.298402Z TRACE isahc::handler] << Cache-Control: no-cache, no-store, max-age=0, must-revalidate\r\n
[2021-02-27T08:37:14.298646Z TRACE isahc::handler] << Content-Encoding: gzip\r\n
[2021-02-27T08:37:14.298802Z TRACE isahc::handler] << Content-Type: application/json\r\n
</pre>

### tcpdump:
<pre>
09:37:11.401841 IP 192.168.178.46.52019 > 52.74.146.247.443: Flags [S], seq 3970471467, win 65535, options [mss 1460,nop,wscale 6,nop,nop,TS val 655569283 ecr 0,sackOK,eol], length 0
09:37:11.717286 IP 52.74.146.247.443 > 192.168.178.46.52019: Flags [S.], seq 2419961259, ack 3970471468, win 26847, options [mss 1420,sackOK,TS val 1787139014 ecr 655569283,nop,<b>wscale 8</b>], length 0
09:37:11.717345 IP 192.168.178.46.52019 > 52.74.146.247.443: Flags [.], ack 1, win 2068, options [nop,nop,TS val 655569595 ecr 1787139014], length 0
...
TLS handshake and other non-pertinent packets
..
09:37:13.573372 IP 52.74.146.247.443 > 192.168.178.46.52019: Flags [P.], seq 5869:5918, ack 1255, <b>win 119</b>, options [nop,nop,TS val 1787140866 ecr 655571127], length 49
09:37:13.573395 IP 192.168.178.46.52019 > 52.74.146.247.443: Flags [.], ack 5869, win 2039, options [nop,nop,TS val 655571442 ecr 1787140866], length 0
09:37:13.573406 IP 192.168.178.46.52019 > 52.74.146.247.443: Flags [.], ack 5918, win 2038, options [nop,nop,TS val 655571442 ecr 1787140866], length 0
<b>09:37:13.583046 IP 192.168.178.46.52019 > 52.74.146.247.443: Flags [P.], seq 1255:1498, ack 5918, win 2048, options [nop,nop,TS val 655571451 ecr 1787140866], length 243
09:37:13.974301 IP 52.74.146.247.443 > 192.168.178.46.52019: Flags [.], ack 1498, win 124, options [nop,nop,TS val 1787141220 ecr 655571451], length 0
09:37:13.974359 IP 192.168.178.46.52019 > 52.74.146.247.443: Flags [P.], seq 1498:2163, ack 5918, win 2048, options [nop,nop,TS val 655571841 ecr 1787141220], length 665</b>
09:37:14.294742 IP 52.74.146.247.443 > 192.168.178.46.52019: Flags [.], ack 2163, win 129, options [nop,nop,TS val 1787141592 ecr 655571841], length 0
09:37:14.297671 IP 52.74.146.247.443 > 192.168.178.46.52019: Flags [P.], seq 5918:6461, ack 2163, win 129, options [nop,nop,TS val 1787141595 ecr 655571841], length 543
</pre>

The timestamps are off by one hour due to the two tools using different timezones. I've bolded the relevant parts of the log that I suspect are causing the increased latency. 

As per `libcurl`, the request headers and body are written together and the whole outbound request has been completed within 1ms (from the message `We are completely uploaded and fine`). Running the client with `strace` shows that both the `write` syscalls from `libcurl` happen close together too so everything seems correct from `curl`s perspective.

However, looking at the `tcpdump` output, the interaction seems to be as below:
1. send headers in one packet
2. wait for `Ack` from remote
3. send request body in next packet

I understand that since `TCP_NODELAY` is on by default in `curl` now, both the headers and body will be sent out in separate packets.  However, why is the kernel waiting for the remote server to `ack` the first packet before sending the second one? 

The remote server had published a `window size` of `119` before this and with a `window scale` of 256, the calculated window size is `30464`. Given the `receive window size` is large enough, I would have expected the kernel to send both the packets back to back. Is this expectation incorrect?

For reference, I observer this behaviour on  `OSX 10.15.7` and `Debian 10.8`.  

Any suggestions on where to look next would be great!

**Reply by tailrecur** [reply]
@bagder Since this issue is not related to `libcurl` anymore, I've moved this to [a question](https://serverfault.com/questions/1055327/kernel-tcp-stack-not-forwarding-packets-immediately-to-remote-server) on ServerFault. 

I'm not familiar with the etiquette around `Github discussions`. Should I mark this as `answered` then?
